yearn/contracts/CurveVoterProxyStrategy.sol

```diff
@@ -0,0 +1,511 @@
+        return crvBalance.sub(crvTransfer);
+    }
+
+    /// @notice This method is defined in the BaseStrategy contract and is meant
+    ///         to perform any strategy unwinding or other calls necessary to
+    ///         capture the free return this strategy has generated since the
+    ///         last time its core position(s) were adjusted. Examples include
+    ///         unwrapping extra rewards. This call is only used during normal
+    ///         operation of a strategy, and should be optimized to minimize
+    ///         losses as much as possible. This strategy implements the
+    ///         aforementioned behavior by harvesting the Curve pool's gauge
+    ///         CRV rewards, using most of the CRV tokens to buy one of the
+    ///         Curve pool's accepted token, and depositing that token back to
+    ///         the Curve pool. This way the strategy is gaining new vault's
+    ///         underlying tokens thus making the profit. A small portion of
+    ///         CRV rewards (defined by keepCRV param) is transferred to the
+    ///         voter contract to increase CRV boost and get more CRV rewards
+    ///         in the future. Apart from that, this strategy also sells
+    ///         the Curve pool's gauge additional rewards, generated using the
+    ///         Synthetix staking rewards contract.
+    /// @param debtOutstanding Will be 0 if the strategy is not past the
+    ///        configured debt limit, otherwise its value will be how far past
+    ///        the debt limit the strategy is. The strategy's debt limit is
+    ///        configured in the vault.
+    /// @return profit Amount of realized profit.
+    /// @return loss Amount of realized loss.
+    /// @return debtPayment Amount of repaid debt. The value of debtPayment
+    ///         should be less than or equal to debtOutstanding. It is okay for
+    ///         it to be less than debtOutstanding, as that should only used as
+    ///         a guide for how much is left to pay back. Payments should be
+    ///         made to minimize loss from slippage, debt, withdrawal fees, etc.
+    function prepareReturn(uint256 debtOutstanding)
+        internal
+        override
+        returns (
+            uint256 profit,
+            uint256 loss,
+            uint256 debtPayment
+        )
+    {
+        // Get the initial balance of the vault's underlying token under
+        // strategy management.
+        uint256 initialWantBalance = want.balanceOf(address(this));
+
+        // Harvest the CRV rewards from the Curve pool's gauge.
+        IStrategyProxy(strategyProxy).harvest(tbtcCurvePoolGauge);
+
+        // Buy WBTC using harvested CRV tokens.
+        uint256 crvBalance = IERC20(crv).balanceOf(address(this));
+        if (crvBalance > 0) {
+            // Deposit a portion of CRV to the voter to gain CRV boost.
+            crvBalance = adjustCRV(crvBalance);
+
+            IERC20(crv).safeApprove(dex, 0);
```

Comment:
why do we need to approve `0` and right after `crvBalance` for the same `crv` token?

Reply:
Please see: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/fa64a1ced0b70ab89073d5d0b6e01b0778f7e7d6/contracts/token/ERC20/IERC20.sol#L42

Reply:
Right, but the link points to `approve`, not [safeApprove](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/fa64a1ced0b70ab89073d5d0b6e01b0778f7e7d6/contracts/token/ERC20/SafeERC20.sol#L37)

Reply:
But `safeApprove` doesn't solve the problem: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/fa64a1ced0b70ab89073d5d0b6e01b0778f7e7d6/contracts/token/ERC20/SafeERC20.sol#L38

Reply:
Indeed. Does `safeIncreaseApproval()` cover it?

Reply:
It looks like [safeIncreaseAllowance()](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v3.4.0/contracts/token/ERC20/SafeERC20.sol#L48) is available in `3.4.0`

Reply:
Good catch! I browsed an older version by mistake 🤦 I'll use it.

Reply:
Here it is: https://github.com/keep-network/tbtc-v2/pull/8/commits/9419c237815fcd015d7c9b766e699ca8cb0bf02f